drivers/uacpi/BindsIo.cpp

```diff
@@ -0,0 +1,226 @@
+#include <uacpi/kernel_api.h>
+#include <interfaces/driver/Memory.h>
+#include <Log.h>
+#include <NativePtr.h>
+#include <Maths.h>
+
+extern "C"
+{
+#ifdef __x86_64__
+    static uacpi_u64 PortRead(uint16_t port, uacpi_u8 width)
+    {
+        union
+        {
+            uint8_t u8;
+            uint16_t u16;
+            uint32_t u32;
+        } temp;
+
+        switch (width)
+        {
+        case 1: asm("inb %1, %0" : "=a"(temp.u8) : "Nd"(port) : "memory"); break;
+        case 2: asm("inw %1, %0" : "=a"(temp.u16) : "Nd"(port) : "memory"); break;
+        case 4: asm("inl %1, %0" : "=a"(temp.u32) : "Nd"(port) : "memory"); break;
+        }
+
+        return temp.u32;
+    }
+
+    static void PortWrite(uint16_t port, uacpi_u8 width, uacpi_u64 data)
+    {        
+        switch (width)
+        {
+        case 1: asm volatile("outb %0, %1" :: "a"((uint8_t)data), "Nd"(port)); break;
+        case 2: asm volatile("outw %0, %1" :: "a"((uint16_t)data), "Nd"(port)); break;
+        case 4: asm volatile("outl %0, %1" :: "a"((uint32_t)data), "Nd"(port)); break;
+        }
+    }
+#endif
+
+    uacpi_status uacpi_kernel_raw_memory_read(uacpi_phys_addr address, uacpi_u8 byte_width, uacpi_u64* out_value)
+    {
+        VALIDATE_(out_value != nullptr, UACPI_STATUS_INVALID_ARGUMENT);
+        VALIDATE_(byte_width <= 8 && sl::IsPowerOfTwo(byte_width), UACPI_STATUS_INVALID_ARGUMENT);
+
+        const size_t hhdmLimit = npk_hhdm_limit();
+        sl::NativePtr access = nullptr;
+
+        if (address < hhdmLimit)
+            access = address + npk_hhdm_base();
+        else
+        {
+            //address is outside of hhdm, we'll need to temporarily map it
+            access = npk_vm_alloc(byte_width, (void*)address, VmMmio, nullptr);
+            if (access.ptr == nullptr)
+                return UACPI_STATUS_INTERNAL_ERROR;
+        }
+
+        switch (byte_width)
+        {
+        case 1: *out_value = access.Read<uint8_t>(); break;
+        case 2: *out_value = access.Read<uint16_t>(); break;
+        case 4: *out_value = access.Read<uint32_t>(); break;
+        case 8: *out_value = access.Read<uint64_t>(); break;
+        }
+
+        if (address >= hhdmLimit)
+            npk_vm_free(access.ptr);
+
+        return UACPI_STATUS_OK;
+    }
+
+    uacpi_status uacpi_kernel_raw_memory_write(uacpi_phys_addr address, uacpi_u8 byte_width, uacpi_u64 in_value)
+    {
+        VALIDATE_(byte_width <= 8 && sl::IsPowerOfTwo(byte_width), UACPI_STATUS_INVALID_ARGUMENT);
+
+        const size_t hhdmLimit = npk_hhdm_limit();
+        sl::NativePtr access = nullptr;
+
+        if (address < hhdmLimit)
```

Comment:
this code is repeated without any change twice, could be more useful to put it in a function somehow?